File: internal/nodes/nodes.go
package nodes

import (
	"go/parser"
	"go/token"
	"strings"

	"github.com/dave/dst"
	"github.com/dave/dst/decorator"
	"github.com/modernice/jotbot/internal/slice"
)

// IsExportedIdentifier determines whether the given identifier is exported in
// Go. An identifier is considered exported if it begins with an uppercase
// letter and is not the blank identifier "_". This function can be used to
// check if an identifier would be accessible from other packages based on Go's
// visibility rules.
func IsExportedIdentifier(identifier string) bool {
	if parts := strings.Split(identifier, ":"); len(parts) > 1 {
		identifier = parts[1]
	}
	if parts := strings.Split(identifier, "."); len(parts) > 1 {
		identifier = parts[1]
	}
	return len(identifier) > 0 && identifier != "_" &&
		strings.ToUpper(identifier[:1]) == identifier[:1]
}

// StripIdentifierPrefix removes any prefix before a colon in the given
// identifier and returns the resulting string without the prefix. If no colon
// is present, it returns the original identifier unchanged.
func StripIdentifierPrefix(identifier string) string {
	if parts := strings.Split(identifier, ":"); len(parts) > 1 {
		return parts[1]
	}
	return identifier
}

// HasDoc reports whether the provided decorations contain any documentation
// comments.
func HasDoc(decs dst.Decorations) bool {
	return len(decs.All()) > 0
}

// Doc extracts the leading comment from the specified node, concatenating all
// lines of the comment into a single string. If removeSlash is true, the "//"
// prefix is removed from each line of the comment before concatenation.
func Doc(n dst.Node, removeSlash bool) string {
	lines := n.Decorations().Start.All()
	if removeSlash {
		lines = slice.Map(lines, trimSlash)
	}
	return strings.Join(lines, "")
}

func trimSlash(s string) string {
	return strings.TrimLeft(strings.TrimPrefix(s, "//"), " ")
}

// Parse reads the given source code and constructs an abstract syntax tree for
// that code. It accepts a string or a byte slice as input and returns a pointer
// to a [*dst.File] representing the parsed code, along with any error
// encountered during parsing. The function also considers comments in the
// source code as part of the parsing process.
func Parse[Code ~string | ~[]byte](code Code) (*dst.File, error) {
	return decorator.ParseFile(token.NewFileSet(), "", code, parser.ParseComments|parser.SkipObjectResolution)
}

// MustParse parses the given code and returns the corresponding DST node. It
// panics if an error occurs during parsing. This convenience function is
// intended for use when the code is guaranteed to be valid and any parse error
// would be considered exceptional. The function accepts a type parameter that
// can be a string or a byte slice, representing the source code to parse. The
// returned DST node can then be used for further manipulation or analysis of
// the parsed code structure.
func MustParse[Code ~string | ~[]byte](code Code) dst.Node {
	node, err := Parse(code)
	if err != nil {
		panic(err)
	}
	return node
}

// Identifier extracts the name and determines the export status of the given
// node. It returns an identifier string with a prefix indicating the kind of
// node, such as "func:", "type:", or "var:", along with a boolean indicating
// whether the identifier is exported.
func Identifier(node dst.Node) (identifier string, exported bool) {
	switch node := node.(type) {
	case *dst.FuncDecl:
		identifier = node.Name.Name

		if node.Recv != nil && len(node.Recv.List) > 0 {
			identifier, _ = methodIdentifier(identifier, node.Recv.List[0].Type)
		}

		identifier = "func:" + identifier
	case *dst.GenDecl:
		if len(node.Specs) == 0 {
			break
		}

		spec := node.Specs[0]

		switch spec := spec.(type) {
		case *dst.TypeSpec:
			identifier = "type:" + spec.Name.Name
		case *dst.ValueSpec:
			identifier = "var:" + spec.Names[0].Name
		}
	case *dst.TypeSpec:
		identifier = "type:" + node.Name.Name
	case *dst.ValueSpec:
		identifier = "var:" + node.Names[0].Name
	}

	return identifier, IsExportedIdentifier(identifier)
}

// Find locates a declaration within the abstract syntax tree of Go code given
// an identifier and the root node. It returns the associated specification, the
// parent node if applicable, and a boolean indicating whether the declaration
// was found. The identifier is expected to be prefixed with a domain specifying
// the type of declaration to search for, such as "func", "type", or "var". This
// function does not traverse nested scopes and is limited to top-level
// declarations or methods on top-level types.
func Find(identifier string, root dst.Node) (dst.Spec, dst.Node, bool) {
	parts := strings.Split(identifier, ":")
	if len(parts) != 2 {
		return nil, nil, false
	}

	switch parts[0] {
	case "func":
		if decl, ok := FindFunc(identifier, root); ok {
			return nil, decl, ok
		}
		if decl, ok := FindInterfaceMethod(identifier, root); ok {
			return nil, decl, ok
		}
		return nil, nil, false
	case "type":
		return FindType(identifier, root)
	case "var":
		return FindValue(identifier, root)
	default:
		return nil, nil, false
	}
}

// FindFunc locates a function declaration within the given root node of the
// abstract syntax tree that matches the specified identifier. It returns the
// function declaration if found and a boolean indicating whether the search was
// successful.
func FindFunc(identifier string, root dst.Node) (fn *dst.FuncDecl, found bool) {
	dst.Inspect(root, func(node dst.Node) bool {
		switch node := node.(type) {
		case *dst.FuncDecl:
			if ident, _ := Identifier(node); ident == identifier {
				fn = node
				found = true
				return false
			}
		}
		return true
	})
	return
}

// FindInterfaceMethod locates a method of a specified interface within the
// given abstract syntax tree node. It returns the method declaration as a
// [*dst.Field] and a boolean indicating whether the method was found. The
// identifier used to specify the method should be in the format
// "interfaceName.methodName". If the method is not found, the returned
// [*dst.Field] will be nil and the boolean will be false.
func FindInterfaceMethod(identifier string, root dst.Node) (method *dst.Field, found bool) {
	parts := strings.Split(identifier, ":")
	if len(parts) == 2 {
		identifier = parts[1]
	}

	parts = strings.Split(identifier, ".")
	if len(parts) != 2 {
		return nil, false
	}

	owner := parts[0]
	met := parts[1]

	dst.Inspect(root, func(node dst.Node) bool {
		switch node := node.(type) {
		case *dst.TypeSpec:
			if node.Name.Name != owner {
				break
			}

			if iface, ok := node.Type.(*dst.InterfaceType); ok {
				for _, field := range iface.Methods.List {
					if len(field.Names) == 0 {
						continue
					}

					if field.Names[0].Name == met {
						method = field
						found = true
						return false
					}
				}
			}
		}
		return true
	})
	return
}

// FindValue locates a variable declaration within the abstract syntax tree
// rooted at the specified node, matching the provided identifier. It returns
// the corresponding value specification, the enclosing general declaration if
// present, and a boolean indicating whether the variable was found.
func FindValue(identifier string, root dst.Node) (spec *dst.ValueSpec, decl *dst.GenDecl, found bool) {
	dst.Inspect(root, func(node dst.Node) bool {
		switch node := node.(type) {
		case *dst.GenDecl:
			if len(node.Specs) == 0 {
				break
			}

			for _, s := range node.Specs {
				switch s := s.(type) {
				case *dst.ValueSpec:
					if ident, _ := Identifier(s); ident == identifier {
						spec = s
						decl = node
						found = true
						return false
					}
				}
			}
		}
		return true
	})
	return
}

// FindType locates a type declaration within the abstract syntax tree of a Go
// source file, given its identifier and the root node of the tree. It returns
// the corresponding type specification, the enclosing general declaration if
// applicable, and a boolean indicating whether the type was found.
func FindType(identifier string, root dst.Node) (spec *dst.TypeSpec, decl *dst.GenDecl, found bool) {
	dst.Inspect(root, func(node dst.Node) bool {
		switch node := node.(type) {
		case *dst.GenDecl:
			for _, s := range node.Specs {
				switch s := s.(type) {
				case *dst.TypeSpec:
					if ident, _ := Identifier(s); ident == identifier {
						spec = s
						decl = node
						found = true
						return false
					}
				}
			}
		}
		return true
	})
	return
}

// CommentTarget determines the appropriate node for attaching a comment within
// a Go abstract syntax tree. It resolves between a specification and its outer
// declaration node to find where a comment should be associated, typically
// returning the node that represents the closest syntactic construct to which
// the comment applies. If no specific association is found, it defaults to
// using the provided outer node.
func CommentTarget(spec dst.Spec, outer dst.Node) dst.Node {
	if spec == nil {
		return outer
	}

	switch spec := spec.(type) {
	case *dst.TypeSpec:
		if decl, ok := outer.(*dst.GenDecl); ok && len(decl.Specs) == 1 {
			return decl
		}
		return spec
	case *dst.ValueSpec:
		if decl, ok := outer.(*dst.GenDecl); ok && len(decl.Specs) == 1 {
			return decl
		}
		return spec
	}

	return outer
}

func methodIdentifier(identifier string, recv dst.Expr) (string, bool) {
	switch recv := recv.(type) {
	case *dst.StarExpr:
		if ident, ok := getIdent(recv.X); ok {
			return "(*" + ident.Name + ")." + identifier, true
		}
	default:
		if ident, ok := getIdent(recv); ok {
			return ident.Name + "." + identifier, true
		}
	}
	return "", false
}

func getIdent(expr dst.Expr) (*dst.Ident, bool) {
	var ident *dst.Ident

	switch e := expr.(type) {
	case *dst.Ident:
		ident = e
	case *dst.IndexListExpr:
		ident, _ = e.X.(*dst.Ident)
	case *dst.IndexExpr:
		ident, _ = e.X.(*dst.Ident)
	}

	if ident == nil {
		return nil, false
	}

	return ident, true
}

File: packages/jotbot/src/nodes.ts
import ts from 'typescript'
import type { Identifier, RawIdentifier } from './identifier'
import { parseIdentifier } from './identifier'

/**
 * Determines whether the provided {@link ts.Node} has any associated leading
 * comments. Returns `true` if comments are present; otherwise, returns `false`.
 */
export function hasComments(node: ts.Node) {
  return !(
    (ts.getLeadingCommentRanges(node.getFullText(), 0) ?? []).length === 0
  )
}

/**
 * Determines whether a given TypeScript {@link ts.Node} is part of an exported
 * declaration. This includes any node that is directly exported or contained
 * within an exported node, but not nodes explicitly marked with the 'export'
 * keyword themselves. This function recursively checks the export status by
 * examining the node's modifiers and traversing up its parent nodes if
 * necessary, until it reaches the top-level {@link ts.SourceFile} or determines
 * the export status. It returns `true` if the node is part of an exported
 * declaration, otherwise `false`.
 */
export function isExported(
  node: ts.Node & { modifiers?: ts.NodeArray<ts.ModifierLike> },
): boolean {
  if (node.kind === ts.SyntaxKind.ExportKeyword) {
    return false
  }

  if (node.modifiers?.some((mod) => mod.kind === ts.SyntaxKind.ExportKeyword)) {
    return true
  }

  if (node.parent && !ts.isSourceFile(node.parent) && !isWithinFunction(node)) {
    return isExported(node.parent)
  }

  return false
}

/**
 * Determines whether the given node is contained within a function. It
 * recursively traverses the parent nodes until it finds a function or reaches
 * the root of the syntax tree. Returns `true` if any parent node is a function
 * as determined by {@link isFunction}, otherwise returns `false`.
 */
export function isWithinFunction(node: ts.Node): boolean {
  if (!node.parent) {
    return false
  }
  if (isFunction(node.parent)) {
    return true
  }
  return isWithinFunction(node.parent)
}

/**
 * Determines if a given {@link ts.Node} represents a variable statement that
 * has been exported. The check is performed by first asserting whether the node
 * is a variable statement, followed by verifying its export status. Returns
 * `true` if the node is a variable statement and is marked as exported;
 * otherwise, returns `false`. This function uses type guards to refine the type
 * of the node if it satisfies the conditions for being an exported variable.
 */
export function isExportedVariable(
  node: ts.Node,
): node is ts.VariableStatement {
  if (!ts.isVariableStatement(node)) {
    return false
  }
  return isExported(node)
}

export type SupportedFunction = ts.FunctionDeclaration | ts.FunctionExpression

/**
 * Determines whether a given {@link ts.Node} is a function, specifically if
 * it's either a {@link ts.FunctionDeclaration} or a {@link
 * ts.FunctionExpression}. Returns true if the node is a function, and false
 * otherwise.
 */
export function isFunction(node: ts.Node): node is SupportedFunction {
  return ts.isFunctionDeclaration(node) || ts.isFunctionExpression(node)
}

/**
 * Determines if a given {@link ts.Node} represents a function that is exported.
 * The node must satisfy the conditions of being a function as defined by {@link
 * isFunction} and being exported as defined by {@link isExported} to be
 * considered an exported function. This function acts as a type guard,
 * narrowing the type of the input node to {@link SupportedFunction} when it
 * returns true.
 */
export function isExportedFunction(node: ts.Node): node is SupportedFunction {
  return isFunction(node) && isExported(node)
}

/**
 * Determines if the provided TypeScript node represents a class that has been
 * exported. This function checks if a given node is a class-like declaration,
 * and if so, verifies whether it is marked with export modifiers directly or
 * indirectly through module exports. It returns true if the class is exported
 * and false otherwise. The result is type-guarded, ensuring that the node can
 * be safely treated as a {@link ts.ClassLikeDeclaration} when the function
 * returns true.
 */
export function isExportedClass(
  node: ts.Node,
): node is ts.ClassLikeDeclaration {
  return isClass(node) && isExported(node)
}

/**
 * Determines whether a given TypeScript node is an interface declaration that
 * has been exported. The function checks if the node represents an interface
 * and, if so, whether it is marked with the export modifier either directly or
 * through module or namespace export patterns. Returns `true` if the node is an
 * exported interface, otherwise `false`. The return type includes a type
 * predicate that refines the input type to {@link ts.InterfaceDeclaration} when
 * the function returns `true`.
 */
export function isExportedInterface(
  node: ts.Node,
): node is ts.InterfaceDeclaration {
  return isInterface(node) && isExported(node)
}

/**
 * Determines whether a given TypeScript node is an {@link
 * ts.InterfaceDeclaration}. Returns `true` if the node is an interface
 * declaration, otherwise `false`.
 */
export function isInterface(node: ts.Node): node is ts.InterfaceDeclaration {
  return ts.isInterfaceDeclaration(node)
}

/**
 * Determines whether a given {@link ts.Node} is a {@link
 * ts.TypeAliasDeclaration}. Returns `true` if the node is a type alias
 * declaration, otherwise `false`.
 */
export function isTypeAlias(node: ts.Node): node is ts.TypeAliasDeclaration {
  return ts.isTypeAliasDeclaration(node)
}

export type SupportedMethod = ts.MethodDeclaration | ts.MethodSignature

/**
 * Determines whether a given TypeScript AST node represents either a method
 * declaration or a method signature. If the node is a method, it narrows the
 * type to {@link SupportedMethod}, which includes both {@link
 * ts.MethodDeclaration} and {@link ts.MethodSignature}. This type predicate can
 * be used to assert the specific kind of method node being dealt with in
 * TypeScript compiler API transformations or analyses.
 */
export function isMethod(node: ts.Node): node is SupportedMethod {
  return ts.isMethodDeclaration(node) || ts.isMethodSignature(node)
}

/**
 * Determines whether a given TypeScript node is private within its context.
 * This includes nodes that are explicitly marked with the `private` or
 * `protected` keyword, as well as properties and methods denoted with the
 * ECMAScript private field syntax (prefixed with `#`). It returns `true` if the
 * node is private or protected, otherwise returns `false`. The node is checked
 * for relevant modifiers, and in the case of properties and methods, their
 * naming conventions to infer privacy.
 */
export function isPrivate<Node extends ts.Node>(node: Node): boolean {
  if (
    (ts.isPropertyDeclaration(node) || ts.isMethodDeclaration(node)) &&
    node.name.getText().startsWith('#')
  ) {
    return true
  }

  let modifiers: ts.Modifier[] = []

  if ('modifiers' in node) {
    modifiers = [...((node?.modifiers as ts.NodeArray<ts.Modifier>) ?? [])]
  }

  return modifiers.some(
    (mod) =>
      mod.kind === ts.SyntaxKind.PrivateKeyword ||
      mod.kind === ts.SyntaxKind.ProtectedKeyword,
  )
}

/**
 * Determines whether a given {@link ts.Node} is a {@link SupportedMethod} that
 * is both public and a member of an exported class. It verifies that the node
 * represents a method, is not private or protected, and belongs to a class that
 * is accessible from outside the module. If these conditions are met, it
 * returns `true`, otherwise `false`.
 */
export function isPublicMethodOfExportedClass(
  node: ts.Node,
): node is SupportedMethod {
  if (!isMethod(node) || isPrivate(node)) {
    return false
  }

  const classNode = getClassOfMethod(node)
  if (!classNode) {
    return false
  }

  return isExported(classNode)
}

/**
 * Returns the class declaration that a method is a part of, if it exists. If
 * the method is not contained within a class, or if the node provided is not a
 * method, this function returns `null`. This function assumes that the input
 * node is already confirmed to be a {@link SupportedMethod}, which includes
 * both method declarations and method signatures.
 */
export function getClassOfMethod(node: SupportedMethod) {
  return node.parent && isClass(node.parent) ? node.parent : null
}

/**
 * Determines whether a given node represents a method that belongs to an
 * exported interface. This function asserts that the node is of type {@link
 * SupportedMethod}, which includes method declarations and method signatures.
 * It ensures the method is part of an interface structure and that the
 * interface is marked for export, making it part of the module's public API
 * surface.
 */
export function isMethodOfExportedInterface(
  node: ts.Node,
): node is SupportedMethod {
  if (!isMethod(node)) {
    return false
  }

  const interfaceNode = getInterfaceOfMethod(node)

  return !!interfaceNode && isExported(interfaceNode)
}

/**
 * Determines if the given node represents a method that is part of a type alias
 * declaration which is exported. If the node is identified as a method and it
 * exists within an exported type alias, the function will return `true`,
 * otherwise `false`. The returned value also asserts the node as a {@link
 * SupportedMethod}, narrowing down its type. This utility is useful for
 * ensuring that methods from exported type aliases are correctly processed or
 * manipulated in accordance with their visibility and usage within a TypeScript
 * module.
 */
export function isMethodOfExportedTypeAlias(
  node: ts.Node,
): node is SupportedMethod {
  if (!isMethod(node)) {
    return false
  }

  const mustBeExported = findParentThatMustBeExported(node)
  if (
    !mustBeExported ||
    mustBeExported.kind !== ts.SyntaxKind.TypeAliasDeclaration
  ) {
    return false
  }

  return !!mustBeExported && isExported(mustBeExported)
}

export type SupportedProperty = ts.PropertyDeclaration | ts.PropertySignature

/**
 * Determines if the given node is a property, either as a declaration or a
 * signature. This is applicable to nodes representing properties in classes,
 * interfaces, and type literals. Returns `true` if the node is a {@link
 * SupportedProperty}, which includes {@link ts.PropertyDeclaration} or {@link
 * ts.PropertySignature}, and `false` otherwise.
 */
export function isProperty(node: ts.Node): node is SupportedProperty {
  return ts.isPropertyDeclaration(node) || ts.isPropertySignature(node)
}

/**
 * Determines if a given node is a {@link SupportedProperty} that is public and
 * belongs to an owner such as a {@link ts.ClassLikeDeclaration}, {@link
 * ts.InterfaceDeclaration}, or {@link ts.TypeAliasDeclaration} which is
 * exported. This function returns `true` if the property is not private and its
 * owner is part of the module's public API, otherwise it returns `false`.
 */
export function isPublicPropertyOfExportedOwner(
  node: ts.Node,
): node is SupportedProperty {
  if (!isProperty(node) || isPrivate(node)) {
    return false
  }

  const mustBeExported = findParentThatMustBeExported(node)

  return !!mustBeExported && isExported(mustBeExported)
}

function findParentThatMustBeExported(node: ts.Node): ts.Node | null {
  if (isClass(node) || isInterface(node) || isTypeAlias(node)) {
    return node
  }

  if (!node.parent) {
    return null
  }

  if (
    ts.isTypeLiteralNode(node.parent) &&
    ts.isTypeAliasDeclaration(node.parent.parent)
  ) {
    return node.parent.parent || null
  }

  return findParentThatMustBeExported(node.parent)
}

/**
 * Determines whether a given TypeScript AST node is an {@link
 * ts.ObjectLiteralExpression}. It returns true if the node is an object literal
 * expression, otherwise false. This type predicate can be used to assert the
 * specific type of a node within the TypeScript compiler API.
 */
export function isLiteralObject(
  node: ts.Node,
): node is ts.ObjectLiteralExpression {
  return ts.isObjectLiteralExpression(node)
}

/**
 * Determines whether a given {@link ts.Node} represents a type alias that is
 * exported. If the node is a type alias and is marked with an export modifier,
 * it returns true, indicating that the type alias is part of the module's
 * public API. Otherwise, it returns false. This function can be useful when
 * generating documentation or analyzing the structure of TypeScript code to
 * understand which types are part of the public interface of a module.
 */
export function isExportedType(node: ts.Node): node is ts.TypeAliasDeclaration {
  return ts.isTypeAliasDeclaration(node) && isExported(node)
}

/**
 * Retrieves the parent {@link ts.InterfaceDeclaration} of a given method if the
 * method is part of an interface. If the method is not contained within an
 * interface, returns `null`. This function is applicable to methods represented
 * by either {@link ts.MethodDeclaration} or {@link ts.MethodSignature}.
 */
export function getInterfaceOfMethod(node: SupportedMethod) {
  return node.parent && isInterface(node.parent) ? node.parent : null
}

/**
 * Determines whether the provided node is a class-like structure. This includes
 * classes defined with the `class` keyword or through variable assignments
 * where the variable's value is a class expression. If true, the function
 * narrows the type to {@link ts.ClassLikeDeclaration}.
 */
export function isClass(node: ts.Node): node is ts.ClassLikeDeclaration {
  return ts.isClassLike(node)
}

/**
 * Locates the declaration and comment target for a specified identifier within
 * the provided abstract syntax tree node. It accepts an identifier which can be
 * a variable, function, class, interface, method, property, or type alias name
 * represented either as a structured {@link Identifier} object or as a raw
 * string {@link RawIdentifier}. If the matching node is found, it returns an
 * object containing the `declaration` and `commentTarget` nodes. Otherwise, it
 * returns `null`. The search includes various checks to ensure the correct type
 * of node is found according to the identifier's type specification.
 */
export function findNode(
  root: ts.Node,
  ident: Identifier | RawIdentifier,
): {
  declaration: ts.Node
  commentTarget: ts.Node
} | null {
  const identifier = typeof ident === 'string' ? parseIdentifier(ident) : ident

  switch (identifier.type) {
    case 'var':
      return findVariable(root, identifier.variableName)
    case 'func':
      return findFunction(root, identifier.functionName)
    case 'class':
      return findClass(root, identifier.className)
    case 'iface':
      return findInterface(root, identifier.interfaceName)
    case 'method':
      return findMethod(root, identifier.ownerName, identifier.methodName)
    case 'prop':
      return findProperty(root, identifier.ownerName, identifier.propertyName)
    case 'type':
      return findType(root, identifier.typeName)
  }
}

/**
 * Locates a variable declaration within the AST, matching the specified
 * variable name, and returns an object containing the declaration node and its
 * parent statement if it is exported. If no matching exported variable is
 * found, it recursively searches child nodes or returns null. This function is
 * designed to be used in scenarios such as documentation generation or analysis
 * tools where it's necessary to find and possibly comment on exported variables
 * given their name. The return type is an object with properties `declaration`
 * of type {@link ts.VariableDeclaration} and `commentTarget` of type {@link
 * ts.VariableStatement}, or `null` if the variable cannot be found.
 */
export function findVariable(
  node: ts.Node,
  variableName: string,
): {
  declaration: ts.VariableDeclaration
  commentTarget: ts.VariableStatement
} | null {
  const traverse = () =>
    ts.forEachChild(node, (node) => findVariable(node, variableName)) ?? null

  if (!ts.isVariableStatement(node)) {
    return traverse()
  }

  if (!isExported(node)) {
    return traverse()
  }

  for (const declaration of node.declarationList.declarations) {
    if (getVariableName(declaration) === variableName) {
      return {
        declaration,
        commentTarget: node,
      }
    }
  }

  return traverse()
}

/**
 * Finds a function declaration in the given TypeScript AST node that matches
 * the specified function name. If the function is found and is exported, it
 * returns an object containing both the declaration and a reference to the node
 * that should be targeted for comments. If no matching exported function is
 * found, it returns `null`. This process includes recursively searching child
 * nodes if necessary. The function must be one of the supported function types,
 * namely a {@link ts.FunctionDeclaration} or a {@link ts.FunctionExpression}.
 */
export function findFunction(
  node: ts.Node,
  functionName: string,
): {
  declaration: SupportedFunction
  commentTarget: SupportedFunction
} | null {
  const traverse = () =>
    ts.forEachChild(node, (node) => findFunction(node, functionName)) ?? null

  if (!isFunction(node)) {
    return traverse()
  }

  if (!isExported(node)) {
    return traverse()
  }

  const name = getFunctionName(node)
  if (name !== functionName) {
    return traverse()
  }

  return {
    declaration: node,
    commentTarget: node,
  }
}

/**
 * Retrieves the name of a given function if available. This function supports
 * both {@link ts.FunctionDeclaration} and {@link ts.FunctionExpression}.
 * Returns the name as a string, or undefined if the function does not have an
 * explicit name.
 */
export function getFunctionName(func: SupportedFunction) {
  if (ts.isFunctionDeclaration(func)) {
    return func.name?.getText()
  }

  if (ts.isFunctionExpression(func)) {
    return func.name?.getText()
  }
}

/**
 * Locates a class declaration with the specified name within the given
 * TypeScript syntax tree node. If a matching exported class is found, it
 * returns an object containing both the class declaration and itself as the
 * target for comments. If no matching class is found, it returns null. This
 * function is useful for tools that need to associate metadata or documentation
 * with specific TypeScript nodes. It ensures that only exported classes are
 * considered, which typically correspond to the API surface of a module or
 * package.
 */
export function findClass(
  node: ts.Node,
  className: string,
): {
  declaration: ts.ClassLikeDeclaration
  commentTarget: ts.ClassLikeDeclaration
} | null {
  if (isClass(node) && isExported(node) && getClassName(node) === className) {
    return {
      declaration: node,
      commentTarget: node,
    }
  }

  return ts.forEachChild(node, (node) => findClass(node, className)) ?? null
}

/**
 * Represents a variable within TypeScript code that is capable of being
 * annotated with comments. The `CommentableVariable` type provides access to
 * both the `statement` part of the variable, which is the entire variable
 * statement, and the `declaration` part, which specifically refers to the
 * variable declaration itself. This dual reference allows for precise targeting
 * when adding or retrieving comments related to a particular variable within
 * the source code.
 */
export interface CommentableVariable {
  statement: ts.VariableStatement
  declaration: ts.VariableDeclaration
}

/**
 * Parses the given variable declaration or statement and returns an object
 * containing both the variable declaration and its associated statement. This
 * function ensures that the variable structure is properly accessed regardless
 * of whether a {@link ts.VariableDeclaration} or {@link ts.VariableStatement}
 * is provided. If a variable statement with no declarations is passed, it
 * throws an error indicating that the expected variable declarations are
 * missing.
 */
export function parseVariable(
  node: ts.VariableDeclaration | ts.VariableStatement,
) {
  if (ts.isVariableDeclaration(node)) {
    return {
      statement: node.parent.parent,
      declaration: node,
    }
  }

  if (!node.declarationList.declarations?.[0])
    throw new Error('Variable statement has no declarations')

  return {
    statement: node,
    declaration: node.declarationList.declarations?.[0] ?? undefined,
  }
}

/**
 * Retrieves the name of a variable from a given TypeScript VariableDeclaration
 * or VariableStatement. If the input is a VariableDeclaration, it extracts the
 * variable's identifier name directly. If the input is a VariableStatement, it
 * identifies the first VariableDeclaration within it and then extracts the name
 * from that declaration. This function is useful for obtaining the string
 * representation of a variable's name for further processing or analysis in
 * code transformation and introspection tasks.
 */
export function getVariableName(
  node: ts.VariableDeclaration | ts.VariableStatement,
) {
  return parseVariable(node).declaration.name.getText()
}

/**
 * Retrieves the name of the owner class, interface, or type alias for a given
 * method or property. If the method or property is not part of a class,
 * interface, or type alias, or if the parent node cannot be determined, the
 * function will return undefined. This function is useful when trying to
 * ascertain the context in which a method or property is defined within the
 * broader structure of TypeScript code. It specifically handles cases where
 * methods and properties are nested within classes, interfaces, and type
 * aliases that are exported from their respective modules.
 */
export function getOwnerName(node: SupportedProperty | SupportedMethod) {
  if (!node.parent) {
    return
  }

  if (isClass(node.parent)) {
    return getClassName(node.parent)
  }

  if (isInterface(node.parent)) {
    return node.parent.name.getText()
  }

  if (ts.isTypeLiteralNode(node.parent) && isTypeAlias(node.parent.parent)) {
    return node.parent.parent.name.getText()
  }
}

/**
 * Retrieves the name of a given class-like declaration. If the class is
 * declared with a name, it returns the textual representation of that name. For
 * anonymous classes, it attempts to infer the name based on its variable or
 * property assignment. If no clear name can be determined, the result is
 * undefined. This function is useful when needing to reference or document the
 * class by its name in situations where a class may not have been explicitly
 * named or when it's being exported as part of a larger expression.
 */
export function getClassName(node: ts.ClassLikeDeclaration) {
  const name = node.name?.getText()
  if (name) {
    return name
  }

  if (!node.parent) {
    return
  }

  if (ts.isVariableDeclaration(node.parent)) {
    return node.parent.name.getText()
  }

  if (ts.isPropertyAssignment(node.parent)) {
    return node.parent.name.getText()
  }
}

/**
 * Finds the specified interface within the provided TypeScript AST node. If the
 * interface with the given name exists, returns an object containing both the
 * interface declaration and the node to which comments can be applied. If no
 * matching interface is found, returns null. This function is useful for tools
 * that need to locate and possibly annotate or modify TypeScript interfaces
 * based on their names within a codebase.
 */
export function findInterface(
  node: ts.Node,
  interfaceName: string,
): {
  declaration: ts.InterfaceDeclaration
  commentTarget: ts.InterfaceDeclaration
} | null {
  if (
    ts.isInterfaceDeclaration(node) &&
    node.name.getText() === interfaceName
  ) {
    return {
      declaration: node,
      commentTarget: node,
    }
  }

  return (
    ts.forEachChild(node, (node) => findInterface(node, interfaceName)) ?? null
  )
}

/**
 * Locates a method within a given TypeScript AST node that matches the
 * specified owner and method names. If found, it returns an object containing
 * the method declaration and the appropriate comment target for documentation
 * purposes, provided the method is public and belongs to either an exported
 * class or interface. Otherwise, returns `null`. The method must not be private
 * or protected to be considered a valid match. This function can be useful when
 * generating documentation or performing code analysis tasks.
 */
export function findMethod(
  node: ts.Node,
  ownerName: string,
  methodName: string,
): {
  declaration: SupportedMethod
  commentTarget: SupportedMethod
} | null {
  const owner = findClass(node, ownerName) ?? findInterface(node, ownerName)
  if (!owner) {
    return null
  }

  const { declaration } = owner

  const members = declaration.members as ts.NodeArray<ts.Node>

  const method = members.find(
    (member) => isMethod(member) && member.name.getText() === methodName,
  )

  if (!method || !isMethod(method)) {
    return null
  }

  if (isPrivate(method)) {
    return null
  }

  return {
    declaration: method,
    commentTarget: method,
  }
}

/**
 * Locates a property with the specified name within the given owner's scope and
 * returns its declaration and the node where comments should be attached. The
 * search is conducted recursively through the AST starting from the provided
 * root node. If the property is found and it is not private, its declaration
 * and comment target are returned; otherwise, null is returned. The owner can
 * be either a class or an interface as determined by the owner name parameter.
 * This function ensures that only publicly accessible properties of exported
 * classes or interfaces are considered in the search.
 */
export function findProperty(
  node: ts.Node,
  ownerName: string,
  propertyName: string,
) {
  const owner = findClass(node, ownerName) ?? findInterface(node, ownerName)
  if (!owner) {
    return null
  }

  const { declaration } = owner

  const members = declaration.members as ts.NodeArray<ts.Node>

  const prop = members.find(
    (member) => isProperty(member) && member.name.getText() === propertyName,
  )

  if (!prop) {
    return null
  }

  if (isPrivate(prop)) {
    return null
  }

  return {
    declaration: prop,
    commentTarget: prop,
  }
}

/**
 * Locates a type alias by its name within the given TypeScript AST node and
 * returns its declaration along with the node that should be targeted for
 * comments. If no matching type alias is found, `null` is returned. The
 * function ensures that the identified type alias has the specified name and
 * that it is part of the AST structure rooted at the provided node. The result
 * includes both the declaration of the type alias itself and the node which
 * should receive associated documentation comments, typically this would be the
 * same as the declaration node. This function is useful for tools that need to
 * associate comments with specific type aliases in a TypeScript codebase.
 * Returns an object containing the `declaration` and `commentTarget` if a match
 * is found, otherwise `null`.
 */
export function findType(
  node: ts.Node,
  typeName: string,
): {
  declaration: ts.TypeAliasDeclaration
  commentTarget: ts.TypeAliasDeclaration
} | null {
  if (ts.isTypeAliasDeclaration(node) && node.name.getText() === typeName) {
    return {
      declaration: node,
      commentTarget: node,
    }
  }

  return ts.forEachChild(node, (node) => findType(node, typeName)) ?? null
}
